modules/conversantAnalyticsAdapter.js

```diff
@@ -0,0 +1,470 @@
+ * AUCTION END event.  Has the following data and key is based off of auctionId, adUnitCode, bidderCode from
+ * keyStr = getLookupKey(auctionId, adUnitCode, bidderCode);
+ * <keyStr> => {
+ *     timeReceived: Date.now() //so cache can be purged in case it doesn't get cleaned out at auctionEnd
+ * }
+ */
+export const timeoutCache = {};
+
+/**
+ * Cleanup timer for the adIdLookup and timeoutCache caches. If all works properly then the caches are self-cleaning
+ * but in case something goes sideways we poll periodically to cleanup old values to prevent a memory leak
+ */
+let cacheCleanupInterval;
+
+let conversantAnalytics = Object.assign(
+  adapter({URL, ANALYTICS_TYPE}),
+  {
+    track({eventType, args}) {
+      if (DO_SAMPLE) {
+        utils.logMessage(eventType, Object.assign({}, args));
+        switch (eventType) {
+          case AUCTION_END:
+            onAuctionEnd(args);
+            break;
+          case AD_RENDER_FAILED:
+            onAdRenderFailed(args);
+            break;
+          case BID_WON:
+            onBidWon(args);
+            break;
+          case BID_TIMEOUT:
+            onBidTimeout(args);
+            break;
+        } // END switch
+      } // END IF(doSample)
+    } // END track()
+  }
+);
+
+// ================================================== EVENT HANDLERS ===================================================
+/**
+ * We get the list of timeouts before the endAution, cache them temporarily in a global cache and the endAuction event
+ * will pick them up.  Uses getLookupKey() to create the key to the entry from auctionId, adUnitCode and bidderCode.
+ * Saves a single value of timeReceived so we can do cache purging periodically.
+ *
+ * Current assumption is that the timeout will always be an array even if it is just one object in the array.
+ * @param args  [{
+  "bidId": "80882409358b8a8",
+    "bidder": "conversant",
+    "adUnitCode": "MedRect",
+    "auctionId": "afbd6e0b-e45b-46ab-87bf-c0bac0cb8881"
+  }, {
+    "bidId": "9da4c107a6f24c8",
+    "bidder": "conversant",
+    "adUnitCode": "Leaderboard",
+    "auctionId": "afbd6e0b-e45b-46ab-87bf-c0bac0cb8881"
+  }
+ ]
+ */
+function onBidTimeout(args) {
+  args.forEach(timedOutBid => {
+    const timeoutCacheKey = getLookupKey(timedOutBid.auctionId, timedOutBid.adUnitCode, timedOutBid.bidder);
+    timeoutCache[timeoutCacheKey] = {
+      timeReceived: Date.now()
+    }
+  });
+}
+
+/**
+ * Bid won occurs after auctionEnd so we need to send this separately. We also save an entry in the adIdLookup cache
+ * so that if the render fails we can match up important data so we can send a valid RENDER FAILED event back.
+ * @param args bidWon args
+ */
+function onBidWon(args) {
+  const bidderCode = args.bidderCode;
+  const adUnitCode = args.adUnitCode;
+  const auctionId = args.auctionId;
+  // Make sure we have all the data we need
+  if (typeof bidderCode === 'undefined' || typeof adUnitCode === 'undefined' || typeof auctionId === 'undefined') {
```

Comment:
These are probably the wrong checks, and will always be true.  Remove 'typeof'.

```javascript
if (bidderCode === 'undefined' || adUnitCode === 'undefined' || auctionId === 'undefined') {
```

Reply:
Nope.  I was wrong.  They work as is.  Without typeof the check should be this instead.

```javascript
if (bidderCode === undefined || adUnitCode === undefined || auctionId === undefined) {
```

Reply:
If you happened to catch some of the deleted comments, this can be simplified further like this:

```javascript
if (!bidderCode || !adUnitCode || !auctionId ) {
```

The upside: catches undefined and null.
The downside: if 0 or false are valid values, then this is not a good check.

Reply:
I'm good with that, I like cleaner code and my way does not account for several of these side cases.  A lot of my checks are based off various stackoverflow threads on "Best practices" which may or may not be outdate.